src/charm.py

```diff
@@ -3,258 +3,112 @@
-    """
 
-    _stored = JenkinsAgentCharmStoredState()
-    service_name = "jenkins-agent"
+class JenkinsAgentCharm(ops.CharmBase):
+    """Charm Jenkins k8s agent."""
 
-    def __init__(self, *args) -> None:
-        """Initialize the instance.
+    def __init__(self, *args: typing.Any):
+        """Initialize the charm and register event handlers.
 
         Args:
-            args: Arguments for the CharmBase superclass.
+            args: Arguments to initialize the charm base.
         """
         super().__init__(*args)
-        self.framework.observe(self.on.start, self._on_config_changed)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.upgrade_charm, self._on_config_changed)
-        self.framework.observe(self.on.slave_relation_joined, self._on_agent_relation_joined)
-        self.framework.observe(self.on.slave_relation_changed, self._on_agent_relation_changed)
+        try:
+            self.state = State.from_charm(self)
+        except InvalidStateError as exc:
+            self.unit.status = ops.BlockedStatus(exc.msg)
+            return
 
-        self._stored.set_default(
-            relation_configured=False,
-            jenkins_url=None,
-            relation_agent_name=f"{self.unit.name.replace('/', '-')}-{uuid.uuid4()}",
-            relation_agent_token=None,
-        )
+        self.pebble_service = pebble.PebbleService(self, self.state)
+        self.agent_observer = agent.Observer(self, self.state, self.pebble_service)
 
-    def _get_pebble_config(self) -> dict:
-        """Generate the pebble config for the charm.
+        self.framework.observe(
+            self.on.jenkins_k8s_agent_pebble_ready, self._register_agent_from_config
+        )
+        self.framework.observe(self.on.config_changed, self._register_agent_from_config)
+        self.framework.observe(self.on.upgrade_charm, self._register_agent_from_config)
 
-        Returns:
-            The pebble configuration for the charm.
-        """
-        env_config = self._get_env_config()
-        pebble_config = {
-            "summary": "jenkins agent layer",
-            "description": "Jenkins Agent layer",
-            "services": {
-                "jenkins-agent": {
-                    "override": "replace",
-                    "summary": "Jenkins Agent service",
-                    "command": "/entrypoint.sh",
-                    "startup": "enabled",
-                    "environment": env_config,
-                }
-            },
-        }
-        return pebble_config
+    @property
+    def _jenkins_agent_container(self) -> ops.Container:
+        """The Jenkins workload container."""
+        return self.unit.get_container(self.state.jenkins_agent_service_name)
 
-    def _on_config_changed(self, event: ConfigChangedEvent) -> None:
-        """Handle config-changed event.
+    def _register_agent_from_config(self, event: ops.HookEvent) -> None:
```

Comment:
I was thinking about whether merging the download + validate part should be merged since it would result in quite a big function that does mostly everything which in my opinion was too much abstraction.
The resulting code would look something like
```
def hook_callback(self: event: ops.HookEvent):
    try:
       creds = server.download_and_validate(server_url, credentials, zip(agent_names, agent_tokens), container, )
    except DownloadFailError:
        # set blocked status
    except ValidateFailError:
        # set Error status for config changed Waiting status for relation changed
```
However, this will make all the handlers very simple and neat. Would this be preferred?

Reply:
I'm a bit hesitant, do we really need the whole logic for both events ? @jdkandersson  what do you think ?